### PR TITLE
feat: Update modules to use new kubectl module

### LINKED ISSUE #604: ASM Module Fails to destroy
Hi,

When trying to destroy a GKE cluster setup with ASM the destroy will fail with:-

```
Error: Error running command 'PATH=<mypath>/.terraform/modules/asm.asm_install/terraform-google-gcloud-1.1.1/cache/94f77b98/google-cloud-sdk/bin:$PATH
.terraform/modules/asm/terraform-google-kubernetes-engine-10.0.0/modules/asm/scripts/kubectl_wrapper.sh https://<IP> <token-removed> kubectl delete ns istio-system
': exit status 1. Output:   -Dd, --decode   decodes input
W0716 13:55:11.407977   22416 loader.go:223] Config not found: /tmp/kubectl_wrapper_9336_20830/config
W0716 13:55:11.408498   22416 loader.go:223] Config not found: /tmp/kubectl_wrapper_9336_20830/config
error: You must be logged in to the server (the server has asked for the client to provide credentials)
```

I know I am logged in locally and can access the k8s API / gCloud etc.

This leaves the terraform state in a bad way and I have to manually delete the resources and the state file.

Any ideas ?

Thanks

### PR DESCRIPTION
**Pull request by bharathkkb**
fixes #604 
- Use kubectl module instead of gcloud wherever possible

- default ACM/ConfigSync modules to `skip_gcloud_download=true`

- TODO
    - [x] new gcloud module release with latest kubectl module

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Error: 
Argument or block definition required
  on terraform-google-kubernetes-engine/autogen/main/variables_defaults.tf line 17:
  (source code not available)
An argument or block definition is required here.
Error: terraform fmt failed with exit code 2
Check the output for diffs and correct using terraform fmt <dir>
Error: 
Argument or block definition required
  on terraform-google-kubernetes-engine/autogen/main/networks.tf line 17:
  (source code not available)
An argument or block definition is required here.
Error: terraform fmt failed with exit code 2
Check the output for diffs and correct using terraform fmt <dir>
 ```

**Comment by morgante** [comment]
@bharathkkb Looks like the comment bot isn't updating?

**Comment by bharathkkb** [comment]
@morgante its actually a bug. The comment bot doesn't seem to ignore the autogen folder and tries to fmt. I will look into this.
image

**Comment by morgante** [comment]
Ok thanks for looking into it.

**Comment by bharathkkb** [comment]
@morgante this should be ready for review

File: modules/asm/main.tf [hunk 1]
```diff
@@ -15,34 +15,26 @@
...
   gcloud_sdk_version                = var.gcloud_sdk_version
   skip_download                     = var.skip_gcloud_download
   upgrade                           = true
   use_tf_google_credentials_env_var = var.use_tf_google_credentials_env_var
-  additional_components             = ["kubectl", "kpt"]
+  additional_components             = ["kubectl", "kpt", "beta", "kustomize"]
 
   create_cmd_entrypoint  = "${path.module}/scripts/install_asm.sh"
   create_cmd_body        = "${var.project_id} ${var.cluster_name} ${var.location}"
   destroy_cmd_entrypoint = "${path.module}/scripts/kubectl_wrapper.sh"
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/asm/main.tf:36]: @morgante the asm installation uses `kpt` and `anthoscli`, not `kubectl`. Hence I have a custom script for that. Only the destroy is a deletion of the istio ns.

File: modules/asm/scripts/kubectl_wrapper.sh [hunk 2]
```diff
@@ -21,33 +21,47 @@ if [ "$#" -lt 3 ]; then
     exit 1
 fi
 
-HOST=$1
-TOKEN=$2
-CA_CERTIFICATE=$3
+CLUSTER_NAME=$1
```
Comment (morgante) [inline modules/asm/scripts/kubectl_wrapper.sh:24]: This specifically looks like duplication of https://github.com/terraform-google-modules/terraform-google-gcloud/blob/master/modules/kubectl-wrapper/scripts/kubectl_wrapper.sh

File: modules/asm/main.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (morgante) [reply modules/asm/main.tf:36]: Shouldn't we still use the submodule though? it should be usable with `anthoscli`.

[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/asm/main.tf:36]: yup, I forgot we exposed the `additional_components` 😅

[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/asm/main.tf:36]: on that note we can replace this too right?
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/4d33759bb6e913586f9d0e2705d6eb2fb6c43a23/autogen/main/dns.tf.tmpl#L22

File: modules/asm/scripts/kubectl_wrapper.sh (hunk 2 above)
Comment (bharathkkb) [reply modules/asm/scripts/kubectl_wrapper.sh]: removed

File: modules/asm/main.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (morgante) [reply modules/asm/main.tf]: Indeed we can.

[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/asm/main.tf]: done

**Comment by bharathkkb** [comment]
@morgante will update once I test, but all these already had a permadiff in every plan due to token refresh.
for example 
- `stub-domains-upstream-nameservers` would always say there are two changes and reapply the shell script which would say `Terraform maintained kube-dns configmap appears to have already been created in kube-system namespace` 
- `acm`, `asm`, `configSync` would always want to do a `destroy_cmd_body ` followed by `create_cmd_body`

**Comment by morgante** [comment]
@bharathkkb You mean they have a permadiff on master or just this branch?

**Comment by bharathkkb** [comment]
@morgante permadiff on master due to new `data.google_client_config.default.access_token` same issue as
https://github.com/terraform-google-modules/terraform-google-gcloud/issues/39#issue-594211060

for instance with `stub-domains-upstream-nameservers` on master after an apply and then the new plan will show
```hcl

  # module.example.module.gke.module.gcloud_delete_default_kube_dns_configmap.null_resource.additional_components[0] must be replaced
-/+ resource "null_resource" "additional_components" {
      ~ id       = "4447235248369243877" -> (known after apply)
      ~ triggers = { # forces replacement
            "additional_components_command" = ".terraform/modules/example.gke.gcloud_delete_default_kube_dns_configmap/terraform-google-gcloud-1.3.0/scripts/check_components.sh gcloud kubectl"
          ~ "arguments"                     = "9f79c50b614be193be9fffe2e3188ca6" -> "c78e810c7855c28edf0ccb2d2e764095"
            "md5"                           = "48135f08d58ee02069d9787f607b6904"
        }
    }

  # module.example.module.gke.module.gcloud_delete_default_kube_dns_configmap.null_resource.run_command[0] must be replaced
-/+ resource "null_resource" "run_command" {
      ~ id       = "1402122050957459867" -> (known after apply)
      ~ triggers = { # forces replacement
          ~ "arguments"              = "9f79c50b614be193be9fffe2e3188ca6" -> "c78e810c7855c28edf0ccb2d2e764095"
          ~ "create_cmd_body"        = "https://123.123.123.123 OLD_TOKEN../../../scripts/delete-default-resource.sh kube-system configmap kube-dns" -> "https://123.123.123.123 ya29.c.NEW_TOKEN../../../scripts/delete-default-resource.sh kube-system configmap kube-dns"
            "create_cmd_entrypoint"  = "../../../scripts/kubectl_wrapper.sh"
            "destroy_cmd_body"       = "info"
            "destroy_cmd_entrypoint" = "gcloud"
            "gcloud_bin_abs_path"    = "/google-cloud-sdk/bin"
            "md5"                    = "48135f08d58ee02069d9787f607b6904"
        }
    }

```

**Comment by morgante** [comment]
@bharathkkb Thanks. Can you test what the upgrade path is too?

**Comment by bharathkkb** [comment]
@morgante  i have tested upgrading some of our examples and documented findings below. I have also confirmed that with this implementation the permadiff issue has been resolved.

- Upgrade path for main module when `stub_domains` or `upstream_nameservers`. Tested with `stub_domains_upstream_nameservers`.
    - No-op as there is no destroy cmd and script checks if `kube-dns` already deleted.
```hcl
 # module.example.module.gke.module.gcloud_delete_default_kube_dns_configmap.null_resource.additional_components[0] will be destroyed
  # module.example.module.gke.module.gcloud_delete_default_kube_dns_configmap.null_resource.additional_components_destroy[0] will be destroyed
  # module.example.module.gke.module.gcloud_delete_default_kube_dns_configmap.null_resource.module_depends_on[0] will be destroyed
  # module.example.module.gke.module.gcloud_delete_default_kube_dns_configmap.null_resource.run_command[0] will be destroyed
  # module.example.module.gke.module.gcloud_delete_default_kube_dns_configmap.module.gcloud_kubectl.null_resource.additional_components[0] will be created
  # module.example.module.gke.module.gcloud_delete_default_kube_dns_configmap.module.gcloud_kubectl.null_resource.additional_components_destroy[0] will be created
  # module.example.module.gke.module.gcloud_delete_default_kube_dns_configmap.module.gcloud_kubectl.null_resource.module_depends_on[0] will be created
  # module.example.module.gke.module.gcloud_delete_default_kube_dns_configmap.module.gcloud_kubectl.null_resource.run_command[0] will be created
```

- Upgrade path for  `acm`. Tested with `simple_zonal_with_acm`.
    - Destroys and reapplies CRDs and manifests _(this is current behavior during every apply due to permadiff on master)_
    - Not pasting whole diff but `13 to add, 0 to change, 37 to destroy`.
    - we also change the default value for `skip_gcloud_download` from `false` to `true`

- Upgrade path for  `asm`. Tested with `simple_zonal_with_asm`.
    - Destroys and reapplies CRDs, manifests _(this is current behavior during every apply due to permadiff on master)_
    - Destroys and recreates hub membership _(this is our example specific as we use had overridden `use_tf_google_credentials_env_var` in the example and `use_tf_google_credentials_env_var` is no longer supported. This can be a breaking change.)_

```hcl
# module.example.module.asm.module.asm_install.null_resource.additional_components[0] will be destroyed
  # module.example.module.asm.module.asm_install.null_resource.additional_components_destroy[0] will be destroyed
  # module.example.module.asm.module.asm_install.null_resource.gcloud_auth_google_credentials[0] will be destroyed
  # module.example.module.asm.module.asm_install.null_resource.gcloud_auth_google_credentials_destroy[0] will be destroyed
  # module.example.module.asm.module.asm_install.null_resource.module_depends_on[0] will be destroyed
  # module.example.module.asm.module.asm_install.null_resource.run_command[0] will be destroyed
  # module.example.module.asm.module.gke_hub_registration.null_resource.gcloud_auth_google_credentials[0] will be destroyed
  # module.example.module.asm.module.gke_hub_registration.null_resource.gcloud_auth_google_credentials_destroy[0] will be destroyed
  # module.example.module.asm.module.gke_hub_registration.null_resource.run_command[0] must be replaced
-/+ resource "null_resource" "run_command" {
  # module.example.module.asm.module.asm_install.module.gcloud_kubectl.null_resource.additional_components[0] will be created
  # module.example.module.asm.module.asm_install.module.gcloud_kubectl.null_resource.additional_components_destroy[0] will be created
  # module.example.module.asm.module.asm_install.module.gcloud_kubectl.null_resource.module_depends_on[0] will be created
  # module.example.module.asm.module.asm_install.module.gcloud_kubectl.null_resource.run_command[0] will be created
```

**Comment by bharathkkb** [comment]
An update for the `acm` example, after swapping to the new module, it did fail initially as the destroy hook was not triggered during the upgrade and hence did not clean up `git-creds` secret. 
I had to do  `.terraform/modules/example.acm.acm_operator.k8sop_creds_secret/terraform-google-gcloud-1.4.0/modules/kubectl-wrapper/scripts/kubectl_wrapper.sh cluster-name us-central1-a project false false kubectl delete secret git-creds -n=config-management-system` to purge it.

**Comment by morgante** [comment]
@bharathkkb Thanks for investigating. Do you think these changes merit a major release?

**Comment by bharathkkb** [comment]
@morgante yeah I believe so because
- we are removing `use_tf_google_credentials_env_var` option from `asm`
- `skip_gcloud_download` defaults are being changed so TF cloud users maybe affected

**Comment by morgante** [comment]
@bharathkkb Sounds good, can you start a small upgrade guide (in separate PR). It should also cover the commands needed to recover/force delete if gcloud has issues.

**Pull request merged**